macos/FluentUI/Button.swift

```diff
@@ -0,0 +1,207 @@
+}
+
+// MARK: - Button
+
+/// A fluent styled button, with hover effects and a corner radius.
+@IBDesignable
+@objc(MSFButton)
+open class Button: NSButton {
+
+	@objc public init(title: String, style: ButtonStyle = .primaryFilled) {
+		super.init(frame: .zero)
+		self.title = title
+		self.style = style
+        initialize()
+    }
+	
+	public override init(frame frameRect: NSRect) {
+        super.init(frame: frameRect)
+        initialize()
+	}
+
+	@available(*, unavailable)
+	required public init?(coder decoder: NSCoder) {
+		preconditionFailure()
+	}
+	
+	open func initialize() {
+		// Do common initialization work
+		isBordered = false
+		wantsLayer = true
+
+		if #available(macOS 10.14, *) {
+			primaryColor = .controlAccentColor
+		} else {
+			primaryColor = .systemBlue
+		}
+		update()
+	}
+	
+	override public func updateTrackingAreas() {
+		super.updateTrackingAreas()
+		
+		// Remove existing trackingArea
+		if let trackingArea = trackingArea {
+			removeTrackingArea(trackingArea)
+			self.trackingArea = nil
+		}
+		
+		// Create a new trackingArea
+		let opts: NSTrackingArea.Options = [.mouseEnteredAndExited, .activeAlways]
+		let trackingArea = NSTrackingArea(
+			rect: bounds,
+			options: opts,
+			owner: self,
+			userInfo: nil
+		)
+		addTrackingArea(trackingArea)
+		self.trackingArea = trackingArea
+	}
+	
+	override public func mouseEntered(with event: NSEvent) {
+		mouseEntered = true
+		needsDisplay = true
+	}
+
+	override public func mouseExited(with event: NSEvent) {
+		mouseEntered = false
+		needsDisplay = true
+	}
+	
+	override public func mouseDown(with event: NSEvent) {
+		mouseDown = true
+		needsDisplay = true
+	}
+	
+	override public func mouseUp(with event: NSEvent) {
+		mouseDown = false
+		needsDisplay = true
+	}
+	
+	open override func updateLayer() {
+		update()
+	}
+	
+	open override var intrinsicContentSize: NSSize {
+		var intrinsicContentSize = super.intrinsicContentSize;
+
+		intrinsicContentSize.width = intrinsicContentSize.width + (horizontalPadding * 2)
+		intrinsicContentSize.height += (verticalPadding * 2);
+
+		return intrinsicContentSize;
+	}
+	
+	override public var wantsUpdateLayer: Bool {
+		return true
+	}
+	
+	/// The primary color of the button, AKA, the fill color in the primaryFilled style, and the outline in the primaryOutline style
+	@objc public var primaryColor: NSColor = .clear {
+		didSet {
+            if primaryColor != oldValue {
+                update()
+            }
+		}
+	}
+
+	/// The primary style of the button
+	@objc public var style: ButtonStyle = .primaryFilled {
+		didSet {
+            if style != oldValue {
+                update()
+            }
+		}
+	}
+	
+	private func update() {
+		if let layer = layer {
+			layer.borderWidth = borderWidth
+			layer.masksToBounds = true
+			layer.cornerRadius = cornerRadius
+			layer.backgroundColor = layerBackgroundColor.cgColor
+			layer.borderColor = outlineColor.cgColor
+		}
```

Comment:
these parts can just go in updateLayer? That would be needed to ensure we properly update from light to dark mode where we use cgColors and not NSColors opaquely